---
title: "single cell basic data processing"
author: "Kent Riemondy RBI"
date: "`R Sys.Date()`"
params:
  ncores: 4
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = FALSE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(cache.lazy = FALSE)
```


```{r libs, message=FALSE, warning=FALSE, echo=FALSE}
source("../R/utils.R")
library(future)

plan("multicore", workers = 7)
options(future.globals.maxSize = 2 * 1024 ^ 3)

proj_dir <- here()
data_dir <- file.path(proj_dir, "data", "cellranger", "results")
doc_dir <- file.path(proj_dir, "docs")

fig_dir <- "figs"
mkrs_dir <- "markers"
tbls_dir <- "tables"
walk(c(fig_dir, mkrs_dir, tbls_dir), dir.create, showWarnings = F)
```

# Experiment Summary

```{r get_data}

# read in metadata to parse out directories and sample ids
mdata <- read_excel(file.path(doc_dir, 
                                   "MED scRNAseq batch and annotations 5.8.19.xlsx")) %>% 

  filter(subgroup != "LGG") %>% 
  select(-subgroup)

mdata_updated <- read_excel(file.path(doc_dir, 
                                   "MED scRNAseq batch and annotations no TBD 5.28.19.xlsx")) %>% 
  rename(subgroup = `subgroup by transcriptomics` , 
         date = ...7,
         fq_id = ...8 ) %>% 
  filter(subgroup != "LGG") %>% 
  select(UPN, subgroup)


mdata <- left_join(mdata_updated, mdata, by = "UPN")

sample_paths <- file.path(data_dir,
                          mdata$fq_id,
                          "outs", 
                          "filtered_feature_bc_matrix")

names(sample_paths) <- str_replace(mdata$UPN, "_", "-")

mat <- Read10X(sample_paths)
```

## General QC for library prep  {.tabset}

```{r }

metrics_paths <- file.path(data_dir,
                          mdata$fq_id,
                          "outs", 
                          "metrics_summary.csv")

names(metrics_paths) <- mdata$UPN

mapping_dat <- map_dfr(metrics_paths, read_csv, .id = "sample")

clean_up_metadata <- function(metrics_summary) {
  metrics_summary <- mutate_all(metrics_summary, str_remove, "%$")
  metrics_summary <- mutate_at(metrics_summary, .vars= 2:ncol(metrics_summary), as.numeric)
  metrics_summary
}

mapping_dat <- clean_up_metadata(mapping_dat)

metrics <- colnames(mapping_dat)[2:ncol(mapping_dat)]

mapping_dat <- mapping_dat %>% 
  gather(metric, value, -sample) %>% 
  left_join(mdata, by = c("sample" = "UPN")) %>% 
  arrange(subgroup) %>% 
  mutate(sample = factor(sample, levels = unique(sample)))

p <- map(metrics, 
    function(x) {
    filter(mapping_dat, metric == x) %>% 
          ggplot(aes(sample, value)) +
            geom_point(aes(color = subgroup)) +
        scale_color_brewer(palette = "Paired") + 
        labs(y = x, 
             x = "") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
}) 
```

```{r, results ='asis'}

for(i in seq_along(p)){
  cat('\n### ', metrics[i], '\n')
  print(p[i])
  cat('\n')
}
```


## Preprocessing {.tabset}


## Rerun preprocessing scaling, PCA, UMAP, etc. 

The dataset alignment approach seems like a poor option for analyzing these tumor samples for the following reason:

1) The normal cells overlap across patients with different tumors suggesting that there isn't a strong global batch effect.
 
2) We already know (or can figure out) which cells are likely malignant based on CNV and expression analysis. We also already know which subgroup they reside in.

3) Clustering the aligned results may give us some insights into gene expression patterns, but really all of the tumor cells will be smashed into a single blob making for a very uninformative visualization 

4) the tumors are heterogenous between patients, so merging them together doesn't seem like a useful way to investigate the heterogeneity in gene expression profiles per tumor. 

Instead of using aligning the data will be presented unaligned. Unlike studies in development or cell-type identificaiton studies, very few visualizations will require UMAPs. The workflow for analysis will be as follows:

1) Project data into UMAP and perform coarse clustering to identify malignant, and non-malignant cells types. 

2) Subset the data into differnet catagories for downstream analysis:
  - 1) normal cells (macrophages, t-cells, b-cells, oligodendrocytes)
    2) subtype specific datasets (SHH and GROUP3/4)
    3) 
    
```{r create_seurat, message = F, results = 'hide', warning = F}

so <- CreateSeuratObject(
  mat,
  min.cells = 3,
  min.features = 200,
  names.delim = "_",
  names.field = 1
)

# fix 966-2 to 966_2 
so@meta.data$orig.ident <- str_replace(so@meta.data$orig.ident,
                                         "966-2", 
                                         "966_2")

so <- PercentageFeatureSet(so, 
                             pattern = "^MT-", 
                             col.name = "percent.mt")

so@meta.data <- so@meta.data %>% 
  tibble::rownames_to_column("cell") %>% 
  left_join(mdata, by = c("orig.ident" = "UPN")) %>% 
  mutate(UPN = orig.ident) %>% 
  as.data.frame() %>% 
  tibble::column_to_rownames("cell")

dir.create("objects", showWarnings = FALSE)

saveRDS(so, file.path("objects", "unfiltered.rds"))
```

### Percent Mitochondrial UMIs 

```{r}
plot_violin(so@meta.data, 
            "orig.ident",
            "percent.mt",
            .fill = "subgroup") +
  labs(x = "", y = "Percent UMIs from Mitochondria")
```


### # of genes detected 

```{r}
plot_violin(so@meta.data,
            "orig.ident", 
            "nFeature_RNA",
            .fill = "subgroup") +
  labs(x = "", y = "# of genes per cell")
```


### # of UMIs detected

```{r}
plot_violin(so@meta.data, 
            "orig.ident",
            "nCount_RNA", 
            .fill = "subgroup") +
  labs(x = "", y = "# of UMIs") 
```

### Table of mitochondrial proportions per sample

```{r}
so@meta.data %>% 
  group_by(UPN) %>% 
  summarize(median_percent_mito = median(percent.mt), 
            mean_percent_mito = mean(percent.mt)) %>% 
  arrange(desc(median_percent_mito))

```


## Relationship between UMIs and % mitochondria {.tabset}

### All samples
```{r, fig.width = 24, fig.height = 18}
sample_names <- as.character(unique(so@meta.data$orig.ident))
per_sample <- map(sample_names, ~filter(so@meta.data, 
                                        orig.ident == .x))
p <- list()
for(i in seq_along(per_sample)){
  .col <- discrete_palette_default[i]
  p[[i]] <- ggplot(per_sample[[i]], aes(nCount_RNA, percent.mt)) +
        geom_point(aes(color = UPN)) +
        scale_color_manual(values = .col)
}

plot_grid(plotlist= p, nrow = 5, ncol = 7)
```

```{r, results ='asis'}

for(i in seq_along(per_sample)){
  .col <- discrete_palette_default[i]
  cat('\n### ', sample_names[i], '\n')
  p <- ggplot(per_sample[[i]], aes(nCount_RNA, percent.mt)) +
        geom_point(aes(color = UPN)) +
        scale_color_manual(values = .col)
  print(p)
  cat('\n')
}
```


## Filter cells and samples.

Suggest removing cells with > 30% mitochondrial UMIs, > 40K UMIs. 

Removing WNT sample from dataset as not enough samples/cells for a rigorous analysis. 

Removing sample 1008 as it only has ~45 malignant cells and is dominated by immune cells

```{r}
so <- readRDS(file.path("objects", "unfiltered.rds"))

so <- subset(so, subset = percent.mt < 30)
so <- subset(so, subset = nCount_RNA < 40000)
so <- subset(so, subset = nCount_RNA > 500)
so <- subset(so, subset = subgroup != "WNT")
so <- subset(so, subset = UPN != "1008")
```


Annotate samples and assign colors to each sample. 


```{r}

so$UPN[so$UPN =="966_2"] <- "966"

annots <- so@meta.data %>% 
  select(UPN, subgroup) %>% 
  unique() %>%
  arrange(subgroup)

subgroup_order <- c(
  "SHH",
  "GP3",
  "GP3/4",
  "GP4"
)

annots <- mutate(annots, subgroup = factor(subgroup, levels = subgroup_order))

annots <- annots %>% arrange(subgroup, as.numeric(UPN))

sample_order <- annots$UPN 

so@meta.data$UPN <- factor(so@meta.data$UPN, levels = sample_order)
so@meta.data$subgroup <- factor(so@meta.data$subgroup, levels = subgroup_order)

# per sample cols
col_pal <- get_distinct_cols(annots$subgroup)

```


## Normalize and embed into 2D with UMAP

```{r}

so <- NormalizeData(so)

so <- FindVariableFeatures(
  so,
  selection.method = "vst",
  nfeatures = 4000,
  verbose = FALSE
)

so <- ScaleData(so, 
                  vars.to.regress = c("percent.mt", "nCount_RNA"), 
                  verbose = TRUE)

so <- RunPCA(so, 
             npcs = 100, 
             verbose = FALSE)

ElbowPlot(so, ndims = 100)

# make graphs and use graphs for UMAP
so <- FindNeighbors(so, 
                      reduction = "pca", 
                      dims = 1:50, 
                      n.neighbors = 50L)


so <- RunUMAP(so, 
              graph = "RNA_snn", 
              min.dist = 0.4)

so <- FindClusters(so, 
                   resolution = c(0.1, 0.3, 0.5))

so$coarse_clusters <- so$RNA_snn_res.0.1

Idents(so) <- "coarse_clusters"

```

```{r}

plot_umap(so, "subgroup", .col = palette_OkabeIto)
plot_umap(so, "coarse_clusters", label_text = TRUE, label_col = "black")
plot_umap(so, "UPN", label_text = TRUE, label_col = "black")
```


```{r}
saveRDS(so, file.path("objects", "so.rds"))
```

## markers 

```{r}
mkrs <- FindAllMarkers(so, only.pos = TRUE)

write_tsv(mkrs, 
          file.path(mkrs_dir, "coarse_cluster_markers_all_data.tsv"))
```



```{r}
so <- readRDS(file.path("objects", "so.rds"))
mkrs <- read_tsv(file.path(mkrs_dir, "coarse_cluster_markers_all_data.tsv"))

topx <- mkrs %>% 
  group_by(cluster) %>% 
  top_n(10, avg_logFC) 
```



```{r, fig.width = 20, fig.height= 20}

p <- DotPlot(so, 
        features = unique(topx$gene),
        group.by = "coarse_clusters") +
  coord_flip() +
  labs(x = "Cluster",
       y = "")

p

save_plot(file.path(fig_dir, 
                    "coarse_cluster_marker_dotplot.pdf"), 
          p, base_height = 24, base_asp = 1)
```


```{r}

cdiv <- so@meta.data %>% 
  rownames_to_column("cell") %>% 
  mutate(n_samples = length(unique(UPN))) %>% 
  group_by(coarse_clusters) %>% 
  mutate(cdiversity = length(unique(UPN)) / n_samples,
         csamples = length(unique(UPN))) %>% 
  ungroup() %>% 
  select(cell, cdiversity, csamples) %>% 
  as.data.frame(.) %>% 
  column_to_rownames("cell")

so <- AddMetaData(so, cdiv)

plot_umap(so, "cdiversity")
```


## Annotate coarse cell types

```{r}
new_ids <- c(
 "0" = "malignant",
 "1" = "malignant",
 "2" = "malignant",
 "3" = "macrophage_monocytes", 
 "4" = "malignant",
 "5" = "malignant",
 "6" = "malignant",
 "7" = "malignant",
 "8" = "malignant",
 "9" = "malignant",
 "10" = "malignant",
 "11" = "lymphocytes", 
 "12" = "malignant",
 "13" = "malignant",
 "14" = "oligodendrocytes_astrocytes_other",
 "15" = "malignant",
 "16" = "malignant"
 )

so$coarse_cell_type <- new_ids[as.character(so$coarse_clusters)]

plot_umap(so, "coarse_cell_type", .cols = palette_OkabeIto)
```



```{r}

to_plot <- c(
  "UPN",
  "coarse_clusters",
  "coarse_cell_type",
  "subgroup",
  "cdiversity",
  "csamples"
) 

pals <- list(
  NULL,
  NULL,
  palette_OkabeIto,
  palette_OkabeIto,
  NULL,
  NULL
)

plts <- map2(to_plot, pals,
    ~plot_umap(so, .x, .cols = .y))

plts


```

## Fix 966 annotation

the two samples from 966 are not replicates, the first is a sample from presentation. The second is from a recurrance sample. 

```{r}
so$UPN <- so$orig.ident

annots <- so@meta.data %>%
  select(UPN, subgroup) %>%
  unique() %>%
  arrange(subgroup)

subgroup_order <- c(
  "SHH",
  "GP3",
  "GP3/4",
  "GP4"
)

annots <- mutate(annots,
                 subgroup = factor(subgroup, levels = subgroup_order))

annots <- annots %>% arrange(subgroup, as.numeric(UPN))

sample_order <- annots$UPN

so@meta.data$UPN <- factor(so@meta.data$UPN, levels = sample_order)

```



## Write out data

```{r} 
saveRDS(so, file.path("objects", "so.rds"))

mdata_out <- get_metadata(so)
outname <- file.path("tables", 
                     paste0("metadata_", format(Sys.Date(), "%Y_%m_%d"), ".tsv.gz"))
write_tsv(mdata_out, outname)
```

