---
title: "Immune populations"
author: "Kent Riemondy RBI"
date: "`r Sys.Date()`"
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "html") })
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libs, message=FALSE, warning=FALSE, echo=FALSE}
source(here::here("/R/utils.R"))

rmd_dir <- "immune"

fig_dir <- file.path(fig_dir, rmd_dir)
mkrs_dir <- file.path(mkrs_dir, rmd_dir)
tbls_dir <- file.path(tbls_dir, rmd_dir)
obj_dir <- file.path(obj_dir, rmd_dir)

walk(c(fig_dir, mkrs_dir, tbls_dir, obj_dir),
     dir.create, showWarnings = F)


seed_value <- 20200204
```

```{r}
so <- qread(file.path("objects", "preprocess", "so.qs"))
```


```{r}
immune_cells <- colnames(so)[so$coarse_cell_type %in% c(
                               "lymphocytes",
                               "macrophage_monocytes")]

so <- so[, immune_cells]
so <- subset(so, subset = subgroup != "WNT")

so <- FindVariableFeatures(so, 
                           selection.method = "vst", 
                           nfeatures = 3000, 
                           verbose = FALSE)

so <- ScaleData(so, 
                vars.to.regress = c("percent.mt", "nCount_RNA"), 
                verbose = TRUE)

so <- RunPCA(so, npcs = 40, verbose = FALSE, seed.use = seed_value)

so <- FindNeighbors(so,
                    reduction = "pca",
                    dims = 1:20,
                    k.param = 20)

so <- RunUMAP(so, graph = "RNA_snn", seed.use = seed_value)


so <- FindClusters(so, 
                   random.seed = seed_value,
                   resolution = c(0.4, 0.3, 0.25, 0.2, 0.1, 0.05))


so$fine_immune_clusters <- so$RNA_snn_res.0.4

clusters <- so$fine_immune_clusters

p <- plot_umap(so, "fine_immune_clusters", label_text = TRUE)
p
save_plot(file.path(fig_dir, "umap_by_clusters.pdf"),
                     p, base_asp = 1.4)

p <- plot_umap(so, "subgroup")
p
save_plot(file.path(fig_dir, "umap_by_subgroup.pdf"), p, base_asp = 1.2)

p <- plot_umap(so, "UPN")
p
save_plot(file.path(fig_dir, "umap_by_upn.pdf"), p, base_asp = 1.2)
```


## Annotate cell types

Using Human primary cell atlas https://www.ncbi.nlm.nih.gov/pubmed/24053356


```{r, fig.height=20, fig.width=8}
library(clustifyr)
hpca_ref <- readRDS("../dbases/human_primary_cell_atlas.rds")

res <- clustify(
  so,
  hpca_ref, 
  cluster_col = "fine_immune_clusters",
  query_genes = so@assays$RNA@var.features,
  seurat_out = FALSE
)

library(ComplexHeatmap)

hmap <- Heatmap(t(res), 
                viridis::viridis(256), 
                "Spearman\ncorrelation",
                row_title = "Cell types from HCPA",
                column_title = "Clusters")
hmap

pdf(file.path(fig_dir, "hmap_hpca.pdf"),
    width = 8, height = 22)
  print(hmap)
dev.off()
```
## add in cluster diversity


```{r}
so <- calc_diversity(so, sample_id = "UPN", group_id = "fine_immune_clusters")
so$cdiversity <- so$entropy
so$entropy <- NULL
p <- plot_umap(so, "cdiversity")
p
save_plot(file.path(fig_dir, "umap_by_cluster_diversity.pdf"), p, base_asp = 1.2)
```

Cluster 
```{r}
mdata <- so@meta.data %>% 
  rownames_to_column("cell")
ctypes  <- call_to_metadata(cor_to_call(res, 
                             metadata = mdata, 
                             cluster_col = "fine_immune_clusters", threshold = 0.5),
                 metadata = mdata,
                 cluster_col = "fine_immune_clusters") %>% 
  mutate(type = ifelse(type == "r<0.5, unassigned",
                       "unknown",
                       type)) %>% 
  pull(type)

# keep broad cell type definition
so$hpca_cell_types <- ctypes %>% 
  str_split(., ":") %>% 
  map_chr(~str_c(.x[1:(length(.x) - 1)], collapse = ":"))

so$hpca_fine_cell_types <- ctypes 

p <- plot_umap(so, "hpca_fine_cell_types")
p
save_plot(file.path(fig_dir, "umap_by_coarse_cell_type.pdf"), p, base_asp = 1.2)

p <- plot_umap(so, "hpca_cell_types")
p
save_plot(file.path(fig_dir, "umap_by_fine_cell_type.pdf"), p, base_asp = 1.2)
```




## Immune cluster markers

```{r}
library(presto)

Idents(so) <- "fine_immune_clusters"

mkrs <- wilcoxauc(so, "fine_immune_clusters")

mkrs <- filter(mkrs, logFC > 0, padj < 0.05) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE)

mkrs %>% 
  write_tsv(file.path(mkrs_dir, "immune_subset_cluster_markers.tsv"))

mkrs %>% 
  ungroup() %>% 
  split(., .$group) %>% 
  write_markers_xlsx(.,
                     file.path(tbls_dir, "immune_subset_cluster_markers.xlsx"))

```

## Cell type composition by subgroup


```{r}
p <- plot_features_split(so, "hpca_cell_types", "subgroup")

p
save_plot(file.path(fig_dir, 
                    "umap_by_fine_cell_types_split_by_subgroup.pdf"), p,
          nrow = 2,
          ncol = 2,
          base_asp = 1.2)
```

```{r}
library(MASC)
library(lme4)

mdata <- so@meta.data

#compare shh to grp3/4
mdata$subgroup_stat_id <- ifelse(mdata$subgroup == "SHH",
                                 "SHH",
                                 "Gp3and4") %>% 
  as.factor()

composition_stats <- MASC:::MASC(data =  mdata, 
            cluster = mdata$fine_immune_clusters, 
            contrast = "subgroup_stat_id",
            random_effects = "UPN") 


composition_stats <- mutate(composition_stats, 
                            log2_OR = log2(subgroup_stat_idSHH.OR),
                            log2_OR_lower = log2(subgroup_stat_idSHH.OR.95pct.ci.lower),
                            log2_OR_upper = log2(subgroup_stat_idSHH.OR.95pct.ci.upper),
                            sig = case_when(
                              model.pvalue < 0.0001 ~ "***",
                              model.pvalue < 0.001 ~ "**",
                              model.pvalue < 0.05 ~ "*",
                              TRUE ~ ""))
                            
composition_stats <- filter(composition_stats, is.finite(subgroup_stat_idSHH.OR.95pct.ci.upper))


p <- ggplot(composition_stats, aes(cluster, log2_OR)) + 
  geom_pointrange(aes(color = -log10(model.pvalue),
                      ymin = log2_OR_lower,
                      ymax = log2_OR_upper)) +
  geom_hline(yintercept = 0, linetype = "dashed") +
  geom_text(aes(label = sig), nudge_y = 1) + 
  scale_color_gradientn(colours = rev(viridis::viridis(256))) +
  coord_flip() +
  labs(y = "Odds ratio (log2)",
       color = "-log10(p)",
       subtitle = "SHH vs. Group3+4")

p

save_plot(file.path(fig_dir, "composition_shift_odds.pdf"), 
          p,
          base_asp = 1.2)
```

```{r}
p <- plot_features_split(so, "fine_immune_clusters", "subgroup")

p

save_plot(file.path(fig_dir, "umap_split_by_cluster.pdf"), 
          p,
          nrow = 2,
          ncol = 2,
          base_asp = 1)

```


## Markers within cluster

```{r}
Idents(so) <- "hpca_cell_types"

so_sub <- subset(so, 
                 subset = hpca_cell_types == "Macrophage:monocyte-derived")

mkrs <- FindMarkers(so_sub, 
                    ident.1 = "SHH", 
                    group.by = "subgroup")

mkrs <- tibble::rownames_to_column(mkrs, "gene")

write_tsv(mkrs, 
          file.path(mkrs_dir, 
                    "macrophage_markers_shh_to_others.tsv"))

mkrs <- set_xlsx_class(mkrs, "gene", "Text")
openxlsx::write.xlsx(mkrs, 
          file.path(tbls_dir, 
                    "macrophage_markers_shh_to_others.xlsx"))

```

```{r}
mkrs <- arrange(mkrs, 
                p_val_adj)
up_mkrs <- filter(mkrs, avg_logFC > 0) %>% slice(1:30)
down_mkrs <- filter(mkrs, avg_logFC < 0) %>% slice(1:30)

top_mkrs <- bind_rows(up_mkrs, down_mkrs)

library(ComplexHeatmap)

Idents(so_sub) <- "subgroup"

to_plot <- log1p(AverageExpression(so_sub)$RNA)
to_plot <- to_plot[top_mkrs$gene, ] 

to_plot <- t(scale(t(as.matrix(to_plot))))

hm <- Heatmap(to_plot, col = viridis::viridis(256),
        name = "Z-score")

pdf(file.path(fig_dir, "hmap_macrophages_differences_between_subgroups.pdf"), height = 12)
print(hm)
dev.off()


```


```{r, eval = FALSE}
so_sub <- subset(so, 
                 subset = refined_immune_clusters == "t-cells")

mkrs <- FindMarkers(so_sub, 
                    ident.1 = "GP4", 
                    ident.2 = "SHH", 
                    group.by = "subgroup")

mkrs <- tibble::rownames_to_column(mkrs, "gene")

write_tsv(mkrs, 
          file.path(mkrs_dir, 
                    "tcells_markers_between_gp4_shh.tsv"))

```

```{r, eval = FALSE}

mkrs <- arrange(mkrs, 
                p_val_adj)
up_mkrs <- filter(mkrs, avg_logFC > 0) %>% slice(1:50)
down_mkrs <- filter(mkrs, avg_logFC < 0) %>% slice(1:50)

top_mkrs <- bind_rows(up_mkrs, down_mkrs)

library(ComplexHeatmap)

Idents(so_sub) <- "subgroup"

to_plot <- log1p(AverageExpression(so_sub)$RNA)
to_plot <- to_plot[top_mkrs$gene, ] 

to_plot <- t(scale(t(as.matrix(to_plot))))

hm <- Heatmap(to_plot, col = viridis::viridis(256),
        name = "Z-score")

pdf("t-cell_differences_between_subgroups.pdf", height = 16)
print(hm)
dev.off()


xcel_out$t_cells <- mkrs

xcel_out <- map(xcel_out, 
                       ~set_xlsx_class(.x, "gene", "Text"))

openxlsx::write.xlsx(xcel_out, 
                     "markers/shh_gp4_immune_cell_differences.xlsx")
```


```{r}
to_plot <- c("refined_immune_clusters", 
"subgroup",
  "UPN")

plts <- map(to_plot, ~plot_umap(so, .x))



plt <- plot_grid(plotlist = plts, nrow = 2, ncol = 2,
                 rel_widths = c(1.75, 1.25))
save_plot("immune_population_umaps.pdf", plt, 
          nrow = 2, ncol = 2, base_asp = 1.2)


plts <- map(c("MRC1", "CCL3"), ~plot_umap(so, .x))


plt <- plot_grid(plotlist = plts, nrow = 1)
save_plot("immune_population_umaps_expr.pdf", plt, 
          nrow = 1, ncol = 2, base_asp = 1.2)

```

```{r}
sc_mdata <- so@meta.data %>% 
  rownames_to_column("cell")

new_annotation <- so@meta.data %>% 
    rownames_to_column("cell") %>% 
    group_by(hpca_cell_types, fine_immune_clusters) %>%
    summarize() %>%
    mutate(fine_cell_type_ids = paste0(hpca_cell_types, "_", "#", row_number())) %>% 
    ungroup() %>% 
  select(fine_immune_clusters, fine_cell_type_ids)

so$fine_cell_type_ids <- left_join(sc_mdata, new_annotation, by = c("fine_immune_clusters")) %>% 
  pull(fine_cell_type_ids)  

sc_mdata <- so@meta.data %>% 
  rownames_to_column("cell")

per_patient <- sc_mdata %>% 
  filter(hpca_cell_types == "Macrophage:monocyte-derived") %>% 
  group_by(UPN) %>%
  mutate(n_cells = n()) %>% 
  group_by(UPN, fine_cell_type_ids) %>% 
  summarize(n = n(),
            prop_cell_type = n / unique(n_cells),
            subgroup = unique(subgroup))

p <- ggplot(per_patient, 
       aes(UPN, prop_cell_type)) +
  geom_col(aes(fill = fine_cell_type_ids)) +
  labs(x = "Tumor ID",
       y = "Proportion of each cell type") +
  scale_fill_manual(values = discrete_palette_default, 
                    name = "") + 
  facet_grid(~subgroup, scales = "free_x", space = "free_x") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        legend.position = "top",
        strip.background = element_rect(fill = "white"))
p
```
## Gene set enrichment

### Macrophages
```{r}
library(gprofiler2)
library(ggrepel)
mkrs <- read_tsv(file.path(mkrs_dir, 
                    "macrophage_markers_shh_to_others.tsv")) %>% 
  filter( p_val_adj < 0.05)

go_res <- gost(mkrs$gene, 
               ordered_query = TRUE,
               sources = c("GO", 
                          "KEGG", 
                          "REAC", 
                          "TF"))
p <- gostplot(go_res, 
              interactive = FALSE)
              

text_data <- group_by(p$data, 
                      source) %>% 
  slice(1:2)

p + 
  geom_text_repel(data = text_data, aes(label = term_name))

go_res[["result"]] %>% 
  mutate(parents = unlist(map(parents, ~str_c(.x, collapse = ",")))) %>% 
  write_tsv(., file.path(tbls_dir, "monocyte_shh_gp4_goterms.tsv"))

go_res[["result"]] %>% 
  mutate(parents = unlist(map(parents, ~str_c(.x, collapse = ","))),
         source = str_replace_all(source, "[[:punct:]]", " ")) %>% 
  select(-significant, -query) %>% 
  split(., .$source) %>% 
  openxlsx::write.xlsx(., file.path(tbls_dir, "monocyte_shh_gp4_goterms.xlsx"))
```


### Each cell type

```{r}
library(gprofiler2)
mkrs <- read_tsv(file.path(mkrs_dir, 
                    "immune_subset_cell_type_markers.tsv")) %>% 
  filter( p_val_adj < 0.05) 

map(unique(mkrs$cluster), function(x){
  clust_mkrs <- filter(mkrs, cluster == x) %>% 
    pull(gene)
go_res <- gost(clust_mkrs, 
               ordered_query = TRUE,
               sources = c("GO", 
                          "KEGG", 
                          "REAC", 
                          "TF"))
p <- gostplot(go_res, 
              interactive = FALSE)
              

text_data <- group_by(p$data, 
                      source) %>% 
  slice(1:2)

p + 
  geom_text_repel(data = text_data, aes(label = term_name))

go_res[["result"]] %>% 
  mutate(parents = unlist(map(parents, ~str_c(.x, collapse = ",")))) %>% 
  write_tsv(., file.path(tbls_dir, str_c(x, "_goterms.tsv")))

go_res[["result"]] %>% 
  mutate(parents = unlist(map(parents, ~str_c(.x, collapse = ","))),
         source = str_replace_all(source, "[[:punct:]]", " ")) %>% 
  select(-significant, -query) %>% 
  split(., .$source) %>% 
  openxlsx::write.xlsx(., file.path(tbls_dir, 
                                    str_c(x, "_goterms.xlsx")))
})
```


## save object

```{r }
qsave(so, file.path(obj_dir, "immune_so.qs"))
```


## Additional annotation


Activate Macrophages look like those seen in HCC (REF) and DAM microglia seen in alzehemiers (Keren-Shaul, H., Spinrad, A., Weiner, A., Matcovitch-Natan, O., DvirSzternfeld, R., Ulland, T.K., David, E., Baruch, K., Lara-Astaiso, D., Toth, B.,
et al. (2017). A unique microglia type associated with restricting development
of Alzheimer’s disease. Cell 169, 1276–1290.e17).

Cluster 4 expresses DC markers (similar to dc markers seen in myleoid cells in CSF of HIV patients, and in the HCC paper)

Cluster 5 is similar to DCs (CD1C+) and Monocytes (CD14+/CD16-) from this scRNA-seq survey: 10.1126/science.aah4573. But also express markers of Neutrophils. 

Cluster 1 is similar to the periperial myeloid cells found in glioblastoma 
(Single-Cell RNA-Seq Analysis of Infiltrating Neoplastic Cells at the Migrating Front of Human Glioblastoma)

```{r}
# immune_new_long_ids <- c(
#   "10" = "Proliferative (MKI67+, TOP2A+)",
#   "11" = "Unknown1 (SEZ6L2, AMER2)",
#   "6" = "Unknown2 (CADM1, OLR1*)",
#   "12" = "Treg (FOXP3+, CTLA4+)",
#   "0" = "T-cells (CD3G+, IL32+)",
#   "7" = "NK-cells (GNLY+, NKG7+)",
#   "9" = "B-cells (CD79A+, IGHM+)",
#  "8" = "Microglia (P2RY12+ TMEM119+)",
#  "3" = "Activated-Microglia 1 (APOE+, GPNMB+)",
#  "2" = "Activated-Microglia 2 (APOE+, TYROBP+)",
#   "1" = "Activated-Microglia 3 (CCL3+, SPP1+)",
#  "5" = "Neutrophil (S100A9+, THBS1+)",
#  "4" = "Dendritic-cell-like (CD1C+,FCER1A+)"
# )
#   
  
immune_new_long_ids <- c(
  "9" = "B-cells (CD79A+, IGHM+)",
  "0" = "T-cells (CD3G+, IL32+)",
  "7" = "NK-cells (GNLY+, NKG7+)",
  "12" = "Treg (FOXP3+, CTLA4+)",
  "10" = "Proliferative (MKI67+, TOP2A+)",
  "5" = "Neutrophil/Monocyte (S100A9+, THBS1+)",
  "8" = "Microglia (P2RY12+ TMEM119+)",
  "6" = "Unknown2 (CADM1, OLR1*)",
  "1" = "TAM M1/M2 (APOE+, TYROBP+)",
  "2" = "TAM CytoHi+ (CCL3+, SPP1+)",
  "3" = "TAM M1/M2 (MRC1+ CD163+)",
  "4" = "Dendritic-cell-like (CD1C+,FCER1A+)",
  "11" = "Unknown1 (SEZ6L2, AMER2)"
)
so@meta.data$new_cell_type_ids <- immune_new_long_ids[as.character(so$fine_immune_clusters)]
   

so@meta.data$new_cell_type_ids_short <- str_split(so$new_cell_type_ids, 
                                                  " ",
                                                  simplify = TRUE) %>% 
  .[, 1]
   
plot_umap(so, "new_cell_type_ids")
plot_umap(so, "new_cell_type_ids_short")
```



Format gene lists for clustifyr


```{r}
library(readxl)
sig_dir <- file.path(proj_dir, "dbases", "gene_signatures")

gene_lsts <- read_excel(file.path(sig_dir, "brca", "mmc4.xlsx"),
                skip = 1) 

gene_lsts <- as.list(gene_lsts) %>% 
  map(~.x[!is.na(.x)])

additional_gene_lsts <- read_excel(file.path(sig_dir, "monocytes", "aah4573_Supplementary_Tables_1-16.xlsx"),
                skip = 2,
                sheet = 2) %>% split(., .$Associated.Cell.Population) %>%
  map(~pull(.x, Gene.ID)) 

gene_lsts <- c(gene_lsts, additional_gene_lsts)

additional_gene_lsts <- read_excel(file.path(sig_dir, "monocytes", "aah4573_Supplementary_Tables_1-16.xlsx"),
                skip = 2,
                sheet = 5) %>% split(., .$Associated.Cell.Population) %>%
  map(~pull(.x, Gene.ID)) 

gene_lsts <- c(gene_lsts, additional_gene_lsts)

# resets length to longest, fills with NA
to_add <- names(gene_lsts)
marker_df <- lapply(gene_lsts, `length<-`, max(lengths(gene_lsts))) %>% 
 map(~tibble(gene = .x)) %>% 
  bind_cols()

colnames(marker_df) <- to_add


res <- clustify_lists(
  input = so,
  per_cell = FALSE,
  marker = marker_df,
  marker_inmatrix = FALSE,
  cluster_col = "fine_immune_clusters",
  seurat_out = FALSE,
  metric = "jaccard",
  dr = "umap"
)
plot_cor_heatmap(res)

# for (i in seq_along(gene_lsts)){
#   so <- AddModuleScore(so, 
#                        features = gene_lsts[i], 
#                        name = names(gene_lsts)[i],
#                        search = TRUE)
# }
# 
# cols_to_plot <- colnames(so@meta.data)[which(colnames(so@meta.data) == "Treg1"):which(colnames(so@meta.data) == "pDC1")]
# 
# map(cols_to_plot, 
#     ~plot_umap(so, .x, show_negative = TRUE))
```


## Average Expression per cluster

```{r}
plot_feature(so, 
             "new_cell_type_ids",
             embedding = "umap")

avg_tbl_dir <- file.path(tbls_dir, "avg_expr", "immune")
dir.create(avg_tbl_dir, recursive = TRUE)

write_avg_expr(so,  
               "new_cell_type_ids",
               file.path(avg_tbl_dir, 
                         "avgexpr.tsv.gz"))
```


```{r}
qsave(so, file.path(obj_dir, "immune_so.qs"))
```

