---
title: "Figures"
author: "Kent Riemondy RBI"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE,
                      warning = FALSE)
```

```{r libs, message=FALSE, warning=FALSE, echo=FALSE}
source("../R/utils.R")
library(future)
plan(strategy = "multicore", workers = 4)
options(future.globals.maxSize = 2 * 1024 ^ 3)

proj_dir <- here()
data_dir <- file.path(proj_dir, "data", "cellranger", "results")
doc_dir <- file.path(proj_dir, "docs")

fig_dir <- "figs"
mkrs_dir <- "markers"
walk(c(fig_dir,mkrs_dir), dir.create, showWarnings = F)
```

```{r get_data}

# read in metadata to parse out directories and sample ids
mdata <- read_excel(file.path(doc_dir, 
                                   "MED scRNAseq batch and annotations 5.8.19.xlsx")) %>% 

  filter(subgroup != "LGG") %>% 
  select(-subgroup)

mdata_updated <- read_excel(file.path(doc_dir, 
                                   "MED scRNAseq batch and annotations no TBD 5.28.19.xlsx")) %>% 
  rename(subgroup = `subgroup by transcriptomics` , 
         date = ...7,
         fq_id = ...8 ) %>% 
  filter(subgroup != "LGG") %>% 
  select(UPN, subgroup)


mdata <- left_join(mdata_updated, mdata, by = "UPN")

sobj <- readRDS(file.path("objects", "unfiltered.rds"))
```

```{r}

annots <- sobj@meta.data %>% 
  select(UPN, subgroup) %>% 
  unique() %>%
  arrange(subgroup)

subgroup_order <- c(
  "WNT",
  "SHH",
  "GP3",
  "GP3/4",
  "GP4"
)

annots <- mutate(annots, 
                 subgroup = factor(subgroup, 
                                   levels = subgroup_order))

annots <- annots %>% 
  arrange(subgroup, as.numeric(UPN))

sample_order <- annots$UPN 

sobj@meta.data$UPN <- factor(sobj@meta.data$UPN, 
                             levels = sample_order)
sobj@meta.data$subgroup <- factor(sobj@meta.data$subgroup, 
                                  levels = subgroup_order)


# per sample cols
#col_pal <- get_distinct_cols(as.character(annots$subgroup))
#names(col_pal) <- annots$UPN
```

## General library statistics  {.tabset}

```{r }

metrics_paths <- file.path(data_dir,
                          mdata$fq_id,
                          "outs", 
                          "metrics_summary.csv")

names(metrics_paths) <- mdata$UPN

mapping_dat <- map_dfr(metrics_paths, read_csv, .id = "sample")

clean_up_metadata <- function(metrics_summary) {
  metrics_summary <- mutate_all(metrics_summary, str_remove, "%$")
  metrics_summary <- mutate_at(metrics_summary, .vars= 2:ncol(metrics_summary), as.numeric)
  metrics_summary
}

mapping_dat <- clean_up_metadata(mapping_dat)

metrics <- colnames(mapping_dat)[2:ncol(mapping_dat)]

mapping_dat <- mapping_dat %>% 
  gather(metric, value, -sample) %>% 
  left_join(mdata, by = c("sample" = "UPN")) %>% 
  arrange(subgroup) %>% 
  mutate(sample = factor(sample, levels = unique(sample)))

p <- map(metrics, 
    function(x) {
    filter(mapping_dat, metric == x) %>% 
          ggplot(aes(sample, value)) +
            geom_point(aes(color = subgroup)) +
        scale_color_brewer(palette = "Paired") + 
        labs(y = x, 
             x = "") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
}) 

```

```{r, results ='asis'}

for(i in seq_along(p)){
  cat('\n### ', metrics[i], '\n')
  print(p[i])
  cat('\n')
}
```



## General QC metrics {.tabset}

### Percent Mitochondrial UMIs 

```{r}
plot_violin(sobj@meta.data, "orig.ident", "percent.mt", .fill = "subgroup") +
  labs(x = "", y = "Percent UMIs from Mitochondria")
```


### # of genes detected 

```{r}
plot_violin(sobj@meta.data, "orig.ident", "nFeature_RNA", .fill = "subgroup") +
  labs(x = "", y = "# of genes per cell")
```


### # of UMIs detected

```{r}
plot_violin(sobj@meta.data, "orig.ident", "nCount_RNA", .fill = "subgroup") +
  labs(x = "", y = "# of UMIs") +
  scale_y_log10()
```

### Table of mitochondrial proportions per sample

```{r}
sobj@meta.data %>% 
  group_by(UPN) %>% 
  summarize(median_percent_mito = median(percent.mt), 
            mean_percent_mito = mean(percent.mt)) %>% 
  arrange(desc(median_percent_mito))

```


## Relationship between UMIs and # of genes {.tabset}

### All samples
```{r, fig.width = 24, fig.height = 18}
sample_names <- levels(sobj@meta.data$UPN)
per_sample <- map(sample_names, ~filter(sobj@meta.data, 
                                        UPN == .x))
p <- list()
for(i in seq_along(per_sample)){
  .col <- discrete_palette_default[i]
  p[[i]] <- ggplot(per_sample[[i]], aes(nFeature_RNA, nCount_RNA)) +
        geom_point(aes(color = UPN)) +
        scale_color_manual(values = .col)
}

plot_grid(plotlist= p, nrow = 5, ncol = 7)
```

```{r, results ='asis'}
# generate tab with individual plot programmatically 
# see https://stackoverflow.com/questions/43752095/programmatically-insert-header-and-plot-in-same-code-chunk-with-r-markdown-using?noredirect=1&lq=1

for(i in seq_along(per_sample)){
  .col <- discrete_palette_default[i]
  cat('\n### ', sample_names[i], '\n')
  p <- ggplot(per_sample[[i]], aes(nFeature_RNA, nCount_RNA)) +
        geom_point(aes(color = UPN)) +
        scale_color_manual(values = .col)
  print(p)
  cat('\n')
}
```



## Relationship between UMIs and % mitochondria {.tabset}

### All samples
```{r, fig.width = 24, fig.height = 18}
sample_names <- levels(sobj@meta.data$UPN)
per_sample <- map(sample_names, ~filter(sobj@meta.data, 
                                        UPN == .x))
p <- list()
for(i in seq_along(per_sample)){
  .col <- discrete_palette_default[i]
  p[[i]] <- ggplot(per_sample[[i]], aes(nCount_RNA, percent.mt)) +
        geom_point(aes(color = UPN)) +
        scale_color_manual(values = .col)
}

plot_grid(plotlist= p, nrow = 5, ncol = 7)
```

```{r, results ='asis'}

# generate tab with individual plot programmatically 
# see https://stackoverflow.com/questions/43752095/programmatically-insert-header-and-plot-in-same-code-chunk-with-r-markdown-using?noredirect=1&lq=1

for(i in seq_along(per_sample)){
  .col <- discrete_palette_default[i]
  cat('\n### ', sample_names[i], '\n')
  p <- ggplot(per_sample[[i]], aes(nCount_RNA, percent.mt)) +
        geom_point(aes(color = UPN)) +
        scale_color_manual(values = .col)
  print(p)
  cat('\n')
}
```





## Filter cells and samples.

I suggest removing cells with > 30% mitochondrial UMIs, > 40K UMIs. Consider also dropping samples with median mito UMIs > 30% These libraries would be `1155`, `996`, `1128`, and `898`. Keeping these libraries for now, but should revisit this point. 

Shown below are the number of cells remaining in each sample and subtype post filtering. There are `r nrow(sobj@meta.data)` total cells in the dataset post filtering. 

```{r}
sobj <- readRDS(file.path("objects", "sobj.rds"))
```


```{r, ncells_post_filtering, fig.caption = "Number of cells per sample remaining after filtering"}
group_by(sobj@meta.data, UPN) %>% 
  summarize(n_cells = n())
```


```{r, ncells_per_subtype_post_filtering, fig.caption = "Number of cells per subgroup remaining after filtering"}
group_by(sobj@meta.data, subgroup) %>% 
  summarize(n_cells = n())
```


## Normalize and embed into 2D with UMAP {.tabset}

All the cells passing initial QC are embedding into 2 dimensions using the UMAP algorithm. Shown below are the 2D projections with cells labeled by either the sample or the subtype of origin. I've additionally plotted a handful of genes, the number of UMIs per cell, and the percentage of reads derived from mitochondrial reads. 

Note that some of the samples were annotated as `TBD`, which I presume means to be determined. 

```{r}
unaligned_dir <- file.path(fig_dir, "unaligned")
dir.create(unaligned_dir, showWarnings = F)
```

### By sample
```{r}
p <- plot_feature(sobj, "UPN", pt_alpha = 0.75, embedding = "umap_unmerged")
p
save_plot(file.path(unaligned_dir,"umap_by_sample.pdf"), p, base_asp = 1.75)
```

### By subtype
```{r}
p <- plot_feature(sobj, "subgroup", pt_alpha = 0.75, embedding = "umap_unmerged")
p
save_plot(file.path(unaligned_dir,"umap_by_subgroup.pdf"), p, base_asp = 1.75)
```


### Various genes

```{r, fig.width = 24, fig.height = 18}
genes <- c(
  'CD3D',
  'PSAP',
  'TOP2A',
  'MYCN',
  'OTX2',
  'CDK6',
  'ACVR1',
  'GFI1B',
  'TERT',
  'SNCAIP',
  'GLI2',
  'YAP1' )

p <- map(genes, ~plot_feature(sobj, .x, embedding = "umap_unmerged"))
p <- plot_grid(plotlist = p, ncol = 4, nrow = 3)

save_plot(file.path(unaligned_dir, "umap_genes.png"), p, 
          base_asp = 1.5, ncol = 4, nrow = 3, dpi = 450)
p
```


### Number of UMIs

```{r}
p <- plot_feature(sobj, "nCount_RNA", embedding = "umap_unmerged")
p
save_plot(file.path(unaligned_dir,"umap_by_nUMI.pdf"), p, base_asp = 1.75)
```

### Percent Mitochondria

```{r}
p <- plot_feature(sobj, "percent.mt", embedding = "umap_unmerged")
p
save_plot(file.path(unaligned_dir,"umap_by_pt_mitochondria.pdf"), p, base_asp = 1.75)
```


## Align data and reembed into 2D with UMAP {.tabset}

Given that many of the samples seem to cluster independently I've also tried using a few dataset alignment methods to improve the clustering. Shown below are new UMAP projections derived from using the Harmony algorithm. 

```{r}
aligned_dir <- file.path(fig_dir, "aligned")
dir.create(aligned_dir, showWarnings = F)
```


### By sample
```{r}
p <- plot_umap(sobj, "UPN", pt_alpha = 0.75)
p
save_plot(file.path(aligned_dir,"umap_by_sample.pdf"), p, base_asp = 1.75)
```

### By subtype

```{r}
p <- plot_umap(sobj, "subgroup", pt_alpha = 0.75)
p
save_plot(file.path(aligned_dir,"umap_by_subgroup.pdf"), p, base_asp = 1.75)
```


### Various genes


```{r, fig.width = 24, fig.height = 18}
genes <- c(
  'CD3D',
  'PSAP',
  'TOP2A',
  'MYCN',
  'OTX2',
  'CDK6',
  'ACVR1',
  'GFI1B',
  'TERT',
  'SNCAIP',
  'GLI2',
  'YAP1' )

p <- map(genes, ~plot_umap(sobj, .x))
p <- plot_grid(plotlist = p, ncol = 4, nrow = 3)
save_plot(file.path(aligned_dir, "umap_genes.png"), p, 
          base_asp = 1.5, ncol = 4, nrow = 3, dpi = 450)
p
```


### Number of UMIs


```{r}
p <- plot_umap(sobj, "nCount_RNA", pt_alpha = 0.75)
p
save_plot(file.path(aligned_dir,"umap_by_nUMI.pdf"), p, base_asp = 1.75)
```

###  Percent UMIs


```{r}
p <- plot_umap(sobj, "percent.mt", pt_alpha = 0.75)
p
save_plot(file.path(aligned_dir,"umap_by_pt_mitochondria.pdf"), p, base_asp = 1.75)
```

## Clustering and finding markers per cluster 

Next I've done some preliminary clustering to try to identify cell types. 


```{r}
p <- plot_umap(sobj, "clusters_aligned", pt_alpha = 0.75)
p
save_plot(file.path(aligned_dir,"umap_by_clusters.pdf"),
          p, base_asp = 1.75)
```

I've generated an excel spreadsheet (found under `markers/markers_per_cluster.xlsx`) with a list of markers for each cluster as a seperate sheet. 

Shown below is dotplot showing the top 5 marker genes for each cluster. 

```{r}
markers <- read_tsv(file.path(mkrs_dir, "aligned_cluster_markers_all_data.tsv"))
topx <- markers %>% 
  group_by(cluster) %>% 
  top_n(5, avg_logFC) 
```

```{r, fig.width = 20, fig.height= 20}

p <- DotPlot(sobj, 
        features = unique(topx$gene),
        group.by = "clusters_aligned") +
  coord_flip() +
  labs(x = "Cluster",
       y = "")

p

save_plot(file.path(aligned_dir, "marker_dotplot.pdf"), 
          p, base_height = 18, base_asp = 1)
```




### compute number of cells per clusters per UPN

```{r}
tbl_dir <- "tables"
x <- dir.create(tbl_dir)

res <- sobj@meta.data %>% 
  group_by(clusters_aligned, UPN) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  spread(UPN, n, fill = 0L) 

res
  write_csv(res, file.path(tbl_dir,
                  "cells_per_cluster_per_sample.csv"))

```


```{r}
tbl_dir <- "tables"
x <- dir.create(tbl_dir)

res <- sobj@meta.data %>% 
  group_by(clusters_aligned, UPN) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  spread(UPN, n, fill = 0L) 

res
  write_csv(res, file.path(tbl_dir,
                  "cells_per_cluster_per_sample.csv"))

res <- sobj@meta.data %>% 
    group_by(clusters_aligned, UPN) %>% 
    summarize(n = n()) %>%
  group_by(clusters_aligned) %>% 
  mutate(total = sum(n),
         percent = 100 * n / total) %>% 
  ungroup() %>% 
  select(-n, -total) %>% 
  spread(UPN, percent, fill = 0L) 

  
res
  write_csv(res, file.path(tbl_dir,
                  "percent_cells_per_cluster_per_sample.csv"))

```



## Annotate cell types

Using Human primary cell atlas https://www.ncbi.nlm.nih.gov/pubmed/24053356


```{r, fig.height=20, fig.width=8}
library(clustifyr)
hpca_ref <- readRDS("extdata/human_primary_cell_atlas.rds")


res <- clustify(
  sobj,
  hpca_ref, 
  cluster_col = "clusters_aligned",
  query_genes = sobj@assays$RNA@var.features,
  seurat_out = FALSE
)

library(ComplexHeatmap)

hmap <- Heatmap(t(res), 
                viridis::viridis(256), 
                "Spearman\ncorrelation",
                row_title = "Cell types from HCPA",
                column_title = "Clusters")
hmap

pdf("figs/aligned/hmap_hpca.pdf", width = 8, height = 20)
print(hmap)
dev.off()
```


Based on Andy's and my analysis cells will be annotated

```{r}

cell_type <- c(
  "17" = "WNT subtype",
  "15" = "WNT outlier_1",
  "16" = "WNT outlier_2",
  "6" = "SHH subtype",
  "4" = "SHH>>GP3>GP4",
"11" = "GP3>>>>GP4",
"7"	= "GP3>>GP4",
"8"	= "GP3>>GP4",
"2"	= "GP3>GP4",
"0"	= "GP4>>GP3",
"1" = "GP4>>GP3",
"14" = "GP4>>GP3",
"18" = "GP3>>GP4",
"13" = "GP4>>>GP3",
"5" = "T-cells",
"3" = "macrophage_monocytes_major",
"12" = "macrophage_monocytes_minor",
"9" = "B-cells",
"10" = "Astrocytes",
"19" = "Oligodendrocytes")


cell_phenotype <- c(
    "17" = "Neoplastic",
  "15" = "Neoplastic",
  "16" = "Neoplastic",
  "6" = "Neoplastic",
  "4" = "Neoplastic",
"11" = "Neoplastic",
"7"	= "Neoplastic",
"8"	= "Neoplastic",
"2"	= "Neoplastic",
"0"	= "Neoplastic",
"1" = "Neoplastic",
"14" = "Neoplastic",
"18" = "Neoplastic",
"13" = "Neoplastic",
"5" = "Immune",
"3" = "Immune",
"12" = "Immune",
"9" = "Immune",
"10" = "Normal",
"19" = "Normal"
)


sobj$cell_type <- cell_type[as.character(sobj$clusters_aligned)]
sobj$cell_phenotype <- cell_phenotype[as.character(sobj$clusters_aligned)]
```


resave object
```{r}
saveRDS(sobj, file.path("objects", "sobj.rds"))
```



