---
title: "Format UCSC cellbrowser"
date: "`r Sys.Date()`"
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "html") })
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Format seurats object into a UCSC cellbrowser. Note requires UCSC (`cellbrowser python package`[ https://cellbrowser.readthedocs.io/index.html]) 

```{r}
source(here::here("/R/utils.R"))
library(scbp)
library(tidyverse)
library(qs)
library(Seurat)
cb_outdir <- "cellbrowser"
alra_objs <- file.path("objects", "alra")
dir.create(alra_objs, recursive = TRUE, showWarnings = FALSE)
def_embeddings <- function(so) {str_subset(names(so@reductions), "umap")}

qread <- function(file, ...) {qs::qread(file, ..., nthreads = 4, use_alt_rep = FALSE)}
```

# All cells


```{r}
so <- qread(file.path("objects", "preprocess", "so.qs"))

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `clusters` = "coarse_clusters",
  `subgroup` = "subgroup",
  `Dx` = "Dx",
  `UPN` = "UPN",
  `other_sample_info` = "additional_info",
   cell_type = "coarse_cell_type",
  `cluster_diversity` = "cdiversity")

make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = c(
                    "subgroup",
                    "cell_type"), 
                 project = "medulloblastoma",
                 outdir = cb_outdir,
                 marker_file = "markers/preprocess/coarse_cluster_markers_all_data.tsv",
                 ident = "clusters",
                 embeddings = def_embeddings(so),
                 skip_expr_matrix = TRUE,
                 config = list(priority = 1,  
                               radius = 2),
                 description = list(    
                   title = "Medulloblastoma",
                   description = "Full dataset, includes all cells."
                 )
)
```


# Malignant only 

```{r}
so <- qread(file.path("objects", "preprocess", "all_neoplastic.qs"))

so$tumor_cell_type_subgroup_specific <- ifelse(is.na( so$tumor_cell_type_subgroup_specific),
                                                "not_classified",
                                                so$tumor_cell_type_subgroup_specific) %>% 
  str_replace_all(., ",", "")

so$tumor_cell_types_all_subgroups <-  str_replace_all(so$tumor_cell_types_all_subgroups, ",", " ")

so$tumor_subpopulation <- ifelse(is.na(so$tumor_subpopulation),
                                 "not-defined",
                                 so$tumor_subpopulation)
cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `unaligned_clusters` = "coarse_clusters_malignant",
  `aligned_clusters` = "coarse_clusters_malignant_harmony",
  `other_sample_info` = "additional_info",
  `subgroup` = "subgroup",
  `subtype` = "subtype",
  `subtype_aligned` = "subtype_aligned",
  `UPN` = "UPN",
  `Cell Cycle Phase` = "Phase",
  `Subgroup specific tumor cell types` = "tumor_cell_type_subgroup_specific",
  `All subgroup tumor cell types` = "tumor_cell_types_all_subgroups",
  `Subgroup subpopulations` = "tumor_subpopulation",
  unaligned_cluster_diversity = "unaligned_cluster_diversity", 
  aligned_cluster_diversity = "aligned_cluster_diversity"
  )

make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = c(
            "subgroup"), 
                 project = "medulloblastoma_neoplastic",
                 outdir = cb_outdir,
                 marker_file = "markers/preprocess/cluster_harmony_markers_neoplastic.tsv",
                 ident = "aligned_clusters",
                 embeddings = def_embeddings(so),
                 skip_expr_matrix = TRUE,
                 config = list(priority = 2,  
                               radius = 2),
                 description = list(    
                   title = "Medulloblastoma Neoplastic cells",
                   description = str_c(c("Only neoplastic cells are shown, includes WNT subgroup. There are both unaligned and aligned projections. Includes cell type labels defined from examining all tumors in a harmonized projection, and cell types defined by harmonizing each subgroup independently.  ",
                   "",
                   "##### Changelog:  ",
                   "2020-04-29 Added Subpopulation designation (e.g. GP3-A1,SHH-B1)  "),
                   collapse = "\n")
                    
                 )
)

gc()
```


# SHH subgroup

```{r}
so <- qread(file.path("objects", "shh", "shh.qs"))

so$tumor_cell_type_shh <- str_replace_all(so$tumor_cell_type_shh, ",", "")
so$tumor_cell_types_all_subgroups <-  str_replace_all(so$tumor_cell_types_all_subgroups, ",", " ")

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `clusters_unaligned` = "coarse_clusters_shh",
  `clusters_aligned` = "coarse_clusters_shh_harmony",
  `subtype` = "subtype",
  `UPN` = "UPN",
  `other_sample_info` = "additional_info",
   cell_type = "coarse_cell_type",
   `cell cycle phase` = "Phase",
  "Northcutt_SHH-A-cell_cycle" = "SHH-A-cell_cycle",
  "Northcutt_SHH-B-SHH_signaling" = "SHH-B-SHH_signaling",
  "Northcutt_SHH-C-Differentiation" = "SHH-C-Differentiation",
  "CytoTRACE-GCS" = "GCS",
  "CytoTRACE-score" = "CytoTRACE",
  unaligned_cluster_diversity = "unaligned_cluster_diversity", 
  aligned_cluster_diversity = "aligned_cluster_diversity",
  `SHH tumor cell types` = "tumor_cell_type_shh",
  `All subgroup tumor cell types` = "tumor_cell_types_all_subgroups",
  `Subpopulations` = "tumor_subpopulations_shh",
  `CNV count (methylation)` = "methyl_nCNV",
  `SHH tumor cell types refined` = "tumor_subpopulations_shh_refined"
  )

so@assays$alra <- get_alra_assay(so,
                                 file.path(alra_objs, "shh.qs"),
                                 overwrite = FALSE)

make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = 
                   c("subtype",
                     "cell_type",
                     "subgroup",
                     "cell cycle phase"), 
                 project = "medulloblastoma_neoplastic_shh",
                 outdir = cb_outdir,
                 marker_file = "markers/shh/harmony_subpopulation_markers_SHH.tsv",
                 ident = "Subpopulations",
                 embeddings = def_embeddings(so),
                 skip_expr_matrix = FALSE, 
                 assays = c("ALRA_" = "alra",
                            "CNV_" = "CNV"),
                  config = list(priority = 3,  
                               radius = 2),
                 description = list(    
                   title = "Medulloblastoma SHH subgroup",
                   description = str_c(c(
                   "Neoplastic cells from SHH subgroup.  ",
                   "There are both unaligned and aligned projections  ",
                   "",
                   "##### Changelog:  ",
                   "2020-05-15 Added 'SHH tumor cell types refined', which contains more refined clustering to produce more gene signatures  ", 
                   "2020-05-15 Added CNV metrics:   ",
                   "",
                   "* CNV count (methylation): # of CNVs based on methylation data  ",
                   "* CNV calls per cell based on inferCNV have been added to the data matrix, which can be viewed by querying the 'Color by gene' tab for cnv metrics. The metrics are prefixed with 'CNV_' (e.g. CNV_has-cnv-chr9, CNV_proportion_cnv-chr22)  ",
                   "  +  The fields has_cnv/has_loss/has_dupli are 0/1 boolean answers to whether respectively any kind of cnv, a loss or a duplication is found within the given chr.  ",
"  +  proportion_cnv/proportion_loss/proportion_dupli is the proportion in number of genes that are part of any cnv/loss cnv/duplication cnv in the given chr.  ",
"  +  proportion_scaled are similar to proportion but taking into account whether genes that are duplications/losses are of a single copy or two copies as a weight  ",
"  +  top_loss_n/top_dupli_n are the n loss/duplication cnvs that cover the most genes  ",
                    "",
                   "2020-04-29 Added Subpopulation designation (e.g. SHH-B1)  "),
                   collapse = "\n")
                 )
)
```


# Gp3 3/4 4 subgroups


```{r}
so <- qread(file.path("objects", "gp34", "gp34.qs"))

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `clusters_unaligned` = "coarse_clusters_gp34",
  `clusters_aligned` = "coarse_clusters_gp34_harmony",
  `subgroup` = "subgroup",
  `subtype` = "subtype",
  `Dx` = "Dx",
  `UPN` = "UPN",
  `other_sample_info` = "additional_info",
   cell_type = "coarse_cell_type",
   `cell cycle phase` = "Phase",
  "Northcutt_Group 3/4-A-cell_cycle" = "Group 3/4-A-cell_cycle",
  "Northcutt_Group 3/4-B-Translation_Myc" = "Group 3/4-B-Translation_Myc",
  "Northcutt_Group 3/4-C-Differentiation" = "Group 3/4-C-Differentiation",
  "CytoTRACE-GCS" = "GCS",
  "CytoTRACE-score" = "CytoTRACE",
  unaligned_cluster_diversity = "unaligned_cluster_diversity", 
  aligned_cluster_diversity = "aligned_cluster_diversity")

#so <- RunALRA(so, setDefaultAssay = FALSE)
make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = 
                   c("subtype",
                     "cell_type",
                     "subgroup",
                     "cell cycle phase"), 
                 project = "medulloblastoma_neoplastic_gp34",
                 outdir = cb_outdir,
                 marker_file = "markers/gp34/harmony_markers_gp34.tsv",
                 ident = "clusters_aligned",
                 embeddings = def_embeddings(so),
                 skip_expr_matrix = TRUE,
                assays = c("ALRA_" = "alra"),
                  config = list(priority = 4),
                 description = list(    
                   title = "Medulloblastoma Group 3, 3+4, and 4 subgroups",
                   description = "Neoplastic cells from group3/4 subgroups.
                   There are both unaligned and aligned projections"
                 )
)
```


# Gp3 subgroup


```{r}
so <- qread(file.path("objects", "gp34", "gp3_only.qs"))

so$tumor_cell_type_gp3 <- str_replace_all(so$tumor_cell_type_gp3, ",", "")
so$tumor_cell_types_all_subgroups <-  str_replace_all(so$tumor_cell_types_all_subgroups, ",", " ")

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `clusters_unaligned` = "coarse_clusters_gp3",
  `clusters_aligned` = "coarse_clusters_gp3_harmony",
  `subgroup` = "subgroup",
  `subtype` = "subtype",
  `Dx` = "Dx",
  `UPN` = "UPN",
    `other_sample_info` = "additional_info",
   `cell cycle phase` = "Phase",
   "Northcutt_Group 3/4-A-cell_cycle" = "Group 3/4-A-cell_cycle",
   "Northcutt_Group 3/4-B-Translation_Myc" = "Group 3/4-B-Translation_Myc",
   "Northcutt_Group 3/4-C-Differentiation" = "Group 3/4-C-Differentiation",
  "CytoTRACE-GCS" = "GCS",
  "CytoTRACE-score" = "CytoTRACE",
  unaligned_cluster_diversity = "unaligned_cluster_diversity", 
  aligned_cluster_diversity = "aligned_cluster_diversity",
  `GP3 tumor cell types` = "tumor_cell_type_gp3",
  `All subgroup tumor cell types` = "tumor_cell_types_all_subgroups",
  `Subpopulations` = "tumor_subpopulations_gp3",
  `CNV count (methylation)` = "methyl_nCNV",
  `GP3 tumor cell types refined` = "tumor_subpopulations_gp3_refined"
  )


so@assays$alra <- get_alra_assay(so,
                                 file.path(alra_objs, "gp3.qs"),
                                 overwrite = FALSE)

scbp::make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = 
                   c("subtype",
                     "cell_type",
                     "subgroup",
                     "cell cycle phase"), 
                 project = "medulloblastoma_neoplastic_gp3_only",
                 outdir = cb_outdir,
                 marker_file = "markers/gp3/harmony_subpopulation_markers_gp3.tsv",
                 ident = "Subpopulations",
                 embeddings = def_embeddings(so),
                 skip_expr_matrix = FALSE,
                 assays = c("ALRA_" = "alra",
                            "CNV_" = "CNV"),
                   config = list(priority = 5),
                 description = list(    
                   title = "Medulloblastoma Group 3 subgroup",
                   description = str_c(c(
                   "Neoplastic cells from group3 subgroup.  ",
                   "There are both unaligned and aligned projections  ",
                   "",
                   "##### Changelog:  ",
                    "2020-05-15 Added 'GP3 tumor cell types refined', which contains more refined clustering to produce more gene signatures  ", 
                   "2020-05-15 Added CNV metrics:   ",
                   "",
                   "* CNV count (methylation): # of CNVs based on methylation data  ",
                   "* CNV calls per cell based on inferCNV have been added to the data matrix, which can be viewed by querying the 'Color by gene' tab for cnv metrics. The metrics are prefixed with 'CNV_' (e.g. CNV_has-cnv-chr9, CNV_proportion_cnv-chr22)  ",
                   "  +  The fields has_cnv/has_loss/has_dupli are 0/1 boolean answers to whether respectively any kind of cnv, a loss or a duplication is found within the given chr.  ",
"  +  proportion_cnv/proportion_loss/proportion_dupli is the proportion in number of genes that are part of any cnv/loss cnv/duplication cnv in the given chr.  ",
"  +  proportion_scaled are similar to proportion but taking into account whether genes that are duplications/losses are of a single copy or two copies as a weight  ",
"  +  top_loss_n/top_dupli_n are the n loss/duplication cnvs that cover the most genes  ",
                    "",
                   "2020-04-29 Added Subpopulation designation (e.g. GP3-B1)  ",
                   "2020-04-29 Reclustered to add in GP3/4 tumor 1167 cells  "),
                   collapse = "\n")
                 )
)
```



# Gp4 subgroup


```{r}
so <- qread(file.path("objects", "gp34", "gp4_only.qs"))

so$tumor_cell_type_gp4 <- str_replace_all(so$tumor_cell_type_gp4, ",", "")
so$tumor_cell_types_all_subgroups <-  str_replace_all(so$tumor_cell_types_all_subgroups, ",", " ")

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `clusters_unaligned` = "coarse_clusters_gp4",
  `clusters_aligned` = "coarse_clusters_gp4_harmony",
  `subgroup` = "subgroup",
  `subtype` = "subtype",
  `Dx` = "Dx",
  `other_sample_info` = "additional_info",
  `UPN` = "UPN",
   `cell cycle phase` = "Phase",
   "Northcutt_Group 3/4-A-cell_cycle" = "Group 3/4-A-cell_cycle",
   "Northcutt_Group 3/4-B-Translation_Myc" = "Group 3/4-B-Translation_Myc",
   "Northcutt_Group 3/4-C-Differentiation" = "Group 3/4-C-Differentiation",
  "CytoTRACE-GCS" = "GCS",
  "CytoTRACE-score" = "CytoTRACE",
  unaligned_cluster_diversity = "unaligned_cluster_diversity", 
  aligned_cluster_diversity = "aligned_cluster_diversity",
  `GP4 tumor cell types` = "tumor_cell_type_gp4",
  `All subgroup tumor cell types` = "tumor_cell_types_all_subgroups",
  `Subpopulations` = "tumor_subpopulations_gp4",
  `CNV count (methylation)` = "methyl_nCNV",
  `GP4 tumor cell types refined` = "tumor_subpopulations_gp4_refined"
)

so@assays$alra <- get_alra_assay(so,
                                 file.path(alra_objs, "gp4.qs"),
                                 overwrite = FALSE)

make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = 
                   c("subtype",
                     "cell_type",
                     "subgroup",
                     "cell cycle phase"), 
                 project = "medulloblastoma_neoplastic_gp4_only",
                 outdir = cb_outdir,
                 marker_file = "markers/gp4/harmony_subpopulation_markers_gp4.tsv",
                 ident = "Subpopulations",
                 embeddings = def_embeddings(so),
                 skip_expr_matrix = FALSE,
                 assays = c("ALRA_" = "alra",
                            "CNV_" = "CNV"),
                config = list(priority = 6),
                 description = list(    
                   title = "Medulloblastoma Group 4 subgroup",
                   description = str_c(c(
                   "Neoplastic cells from group4 subgroup.  ",
                   "There are both unaligned and aligned projections  ",
                   "",
                   "##### Changelog:  ",
                   "2020-05-15 Added 'GP4 tumor cell types refined', which contains more refined clustering to produce more gene signatures  ", 
                   "2020-05-15 Added CNV metrics:   ",
                   "",
                   "* CNV count (methylation): # of CNVs based on methylation data  ",
                   "* CNV calls per cell based on inferCNV have been added to the data matrix, which can be viewed by querying the 'Color by gene' tab for cnv metrics. The metrics are prefixed with 'CNV_' (e.g. CNV_has-cnv-chr9, CNV_proportion_cnv-chr22)  ",
                   "  +  The fields has_cnv/has_loss/has_dupli are 0/1 boolean answers to whether respectively any kind of cnv, a loss or a duplication is found within the given chr.  ",
"  +  proportion_cnv/proportion_loss/proportion_dupli is the proportion in number of genes that are part of any cnv/loss cnv/duplication cnv in the given chr.  ",
"  +  proportion_scaled are similar to proportion but taking into account whether genes that are duplications/losses are of a single copy or two copies as a weight  ",
"  +  top_loss_n/top_dupli_n are the n loss/duplication cnvs that cover the most genes  ",
                    "",
                   "2020-04-29 Added Subpopulation designation (e.g. GP4-B1)  "),
                   collapse = "\n")
                 )
)
```



# Immune populations



```{r}
so <- qread(file.path("objects", "immune", "immune_so.qs"))

# dont use commas in cluster ids
so$new_cell_type_ids <- str_remove(so$new_cell_type_ids, ",")

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `clusters` = "fine_immune_clusters",
  `subgroup` = "subgroup",
  `other_sample_info` = "additional_info",
  `UPN` = "UPN",
 # cell_type_fine = "fine_cell_type_ids",
  new_cell_type_fine = "new_cell_type_ids",
 new_cell_type_short_label = "new_cell_type_ids_short",
  hpca_cell_type_coarse = "hpca_cell_types",
  hpca_cell_type_fine = "hpca_fine_cell_types")

#so <- RunALRA(so,  setDefaultAssay = FALSE)

make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = c(
                   "hpca_cell_type_coarse",
                   "subgroup"), 
                 project = "medulloblastoma_immune",
                 outdir = cb_outdir,
                 marker_file = "markers/immune/immune_subset_cluster_markers.tsv",
                 ident = "clusters",
                 embeddings = def_embeddings(so),
                 skip_expr_matrix = TRUE,
                 assays = c("ALRA_" = "alra"),
                 config = list(priority = 7),
                 description = list(    
                   title = "Medulloblastoma immune populations",
                   description = "Myleoid and lymphocytes populations from all subgroups.
                   There is only an unaligned projection, as alignment was not necessary"
                 )
)
```


# Mouse model (SHH)

```{r, eval = FALSE}
so <- readRDS( file.path("objects",
                  "mouse_models",
                  "shh_model_sobj.rds"))

signatures <- paste0("SHH_human_signatures_clust_", 0:7)
names(signatures) <- signatures

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `clusters` = "clusters",
  `sample` = "orig.ident",
   cell_type = "cell_types",
  `cell cycle phase` = "Phase",
  "Northcutt_MetaProgam_A" = "SHH-A_metafactor1",
  "Northcutt_MetaProgam_B" = "SHH-B_metafactor1",
  "Northcutt_MetaProgam_C" = "SHH-C_metafactor1")

cols_to_keep <- c(cols_to_keep, signatures)
make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = c(
            "sample",
            "cell_type",
            "cell cycle phase"), 
                 project = "medulloblastoma_shh_mouse_model",
                 outdir = cb_outdir,
                 marker_file = "markers/mouse_models/tsv/shh_cluster_markers.tsv",
                 ident = "clusters",
            embeddings = def_embeddings(so),
                 skip_expr_matrix = TRUE
)
```

# Mouse model (Gp34)

```{r, eval = FALSE}
so <- readRDS(file.path("objects",
                  "mouse_models",
                  "mb_models_sobj.rds"))

signatures <- paste0("gp34_human_signatures_clust_",
                     0:7)
names(signatures) <- signatures

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `clusters` = "clusters",
  `sample` = "expt_ids",
   cell_type = "cell_types",
   `cell cycle phase` = "Phase",
   firefly_expression = "luc-firefly",
   egfp_expression = "egfp",
  "Northcutt_MetaProgam_A" = "Group 3/4-A_metafactor1",
  "Northcutt_MetaProgam_B" = "Group 3/4-B_metafactor1",
  "Northcutt_MetaProgam_C" = "Group 3/4-C_metafactor1"
  )

cols_to_keep <- c(cols_to_keep, signatures)


make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = c(
                    "sample",
                    "cell_type",
                    "cell cycle phase"), 
                 project = "medulloblastoma_gp34_mouse_models",
                 outdir = cb_outdir,
                 marker_file = "markers/mouse_models/tsv/mb_models_cluster_markers.tsv",
                 ident = "clusters",
                 embeddings = def_embeddings(so),
                 skip_expr_matrix = TRUE
                 )
```

# Build all 

```{r}
datasets <- c(
   "medulloblastoma/cellbrowser.conf",
   "medulloblastoma_neoplastic/cellbrowser.conf",
   "medulloblastoma_neoplastic_shh/cellbrowser.conf",
   "medulloblastoma_neoplastic_gp34/cellbrowser.conf",
   "medulloblastoma_immune/cellbrowser.conf",
   "medulloblastoma_neoplastic_gp3_only/cellbrowser.conf",
   "medulloblastoma_neoplastic_gp4_only/cellbrowser.conf"
)

datasets <- file.path(cb_outdir, datasets)

build_cellbrowser(datasets, outdir = file.path(cb_outdir, "mblast"))

```




