---
title: "single cell basic data processing"
author: "Kent Riemondy RBI"
date: "`R Sys.Date()`"
params:
  ncores: 4
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = FALSE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(cache.lazy = FALSE)
```


```{r libs, message=FALSE, warning=FALSE, echo=FALSE}
source("../R/utils.R")
library(future)
message(paste0("using ", params$ncores))

plan("multicore", workers = as.numeric(params$ncores))
options(future.globals.maxSize = 2 * 1024 ^ 3)

proj_dir <- here()
data_dir <- file.path(proj_dir, "data", "cellranger", "results")
doc_dir <- file.path(proj_dir, "docs")

fig_dir <- "figs"
mkrs_dir <- "markers"
```

# Experiment Summary

```{r get_data}

# read in metadata to parse out directories and sample ids
mdata <- read_excel(file.path(doc_dir, 
                                   "MED scRNAseq batch and annotations 5.8.19.xlsx")) %>% 

  filter(subgroup != "LGG") %>% 
  select(-subgroup)

mdata_updated <- read_excel(file.path(doc_dir, 
                                   "MED scRNAseq batch and annotations no TBD 5.28.19.xlsx")) %>% 
  rename(subgroup = `subgroup by transcriptomics` , 
         date = ...7,
         fq_id = ...8 ) %>% 
  filter(subgroup != "LGG") %>% 
  select(UPN, subgroup)


mdata <- left_join(mdata_updated, mdata, by = "UPN")

sample_paths <- file.path(data_dir,
                          mdata$fq_id,
                          "outs", 
                          "filtered_feature_bc_matrix")

names(sample_paths) <- str_replace(mdata$UPN, "_", "-")

mat <- Read10X(sample_paths)
```

## General QC for library prep  {.tabset}

```{r }

metrics_paths <- file.path(data_dir,
                          mdata$fq_id,
                          "outs", 
                          "metrics_summary.csv")

names(metrics_paths) <- mdata$UPN

mapping_dat <- map_dfr(metrics_paths, read_csv, .id = "sample")

clean_up_metadata <- function(metrics_summary) {
  metrics_summary <- mutate_all(metrics_summary, str_remove, "%$")
  metrics_summary <- mutate_at(metrics_summary, .vars= 2:ncol(metrics_summary), as.numeric)
  metrics_summary
}

mapping_dat <- clean_up_metadata(mapping_dat)

metrics <- colnames(mapping_dat)[2:ncol(mapping_dat)]

mapping_dat <- mapping_dat %>% 
  gather(metric, value, -sample) %>% 
  left_join(mdata, by = c("sample" = "UPN")) %>% 
  arrange(subgroup) %>% 
  mutate(sample = factor(sample, levels = unique(sample)))

p <- map(metrics, 
    function(x) {
    filter(mapping_dat, metric == x) %>% 
          ggplot(aes(sample, value)) +
            geom_point(aes(color = subgroup)) +
        scale_color_brewer(palette = "Paired") + 
        labs(y = x, 
             x = "") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
}) 
```

```{r, results ='asis'}

for(i in seq_along(p)){
  cat('\n### ', metrics[i], '\n')
  print(p[i])
  cat('\n')
}
```


## Preprocessing {.tabset}

```{r create_seurat, message = F, results = 'hide', warning = F}

sobj <- CreateSeuratObject(mat, 
                             min.cells = 3, 
                             min.features = 200,
                             names.delim = "_", 
                             names.field = 1)

# fix 966-2 to 966_2 
sobj@meta.data$orig.ident <- str_replace(sobj@meta.data$orig.ident, "966-2", "966_2")

sobj <- PercentageFeatureSet(sobj, pattern = "^MT-", col.name = "percent.mt")

sobj@meta.data <- sobj@meta.data %>% 
  tibble::rownames_to_column("cell") %>% 
  left_join(mdata, by = c("orig.ident" = "UPN")) %>% 
  mutate(UPN = orig.ident) %>% 
  as.data.frame() %>% 
  tibble::column_to_rownames("cell")

dir.create("objects", showWarnings = FALSE)

saveRDS(sobj, file.path("objects", "unfiltered.rds"))
```

### Percent Mitochondrial UMIs 

```{r}
plot_violin(sobj@meta.data, "orig.ident", "percent.mt", .fill = "subgroup") +
  labs(x = "", y = "Percent UMIs from Mitochondria")
```


### # of genes detected 

```{r}
plot_violin(sobj@meta.data, "orig.ident", "nFeature_RNA", .fill = "subgroup") +
  labs(x = "", y = "# of genes per cell")
```


### # of UMIs detected

```{r}
plot_violin(sobj@meta.data, "orig.ident", "nCount_RNA", .fill = "subgroup") +
  labs(x = "", y = "# of UMIs") 
```

### Table of mitochondrial proportions per sample

```{r}
sobj@meta.data %>% 
  group_by(UPN) %>% 
  summarize(median_percent_mito = median(percent.mt), 
            mean_percent_mito = mean(percent.mt)) %>% 
  arrange(desc(median_percent_mito))

```


## Relationship between UMIs and % mitochondria {.tabset}

### All samples
```{r, fig.width = 24, fig.height = 18}
sample_names <- as.character(unique(sobj@meta.data$orig.ident))
per_sample <- map(sample_names, ~filter(sobj@meta.data, 
                                        orig.ident == .x))
p <- list()
for(i in seq_along(per_sample)){
  .col <- discrete_palette_default[i]
  p[[i]] <- ggplot(per_sample[[i]], aes(nCount_RNA, percent.mt)) +
        geom_point(aes(color = UPN)) +
        scale_color_manual(values = .col)
}

plot_grid(plotlist= p, nrow = 5, ncol = 7)
```

```{r, results ='asis'}



# generate tab with individual plot programmatically 
# see https://stackoverflow.com/questions/43752095/programmatically-insert-header-and-plot-in-same-code-chunk-with-r-markdown-using?noredirect=1&lq=1

for(i in seq_along(per_sample)){
  .col <- discrete_palette_default[i]
  cat('\n### ', sample_names[i], '\n')
  p <- ggplot(per_sample[[i]], aes(nCount_RNA, percent.mt)) +
        geom_point(aes(color = UPN)) +
        scale_color_manual(values = .col)
  print(p)
  cat('\n')
}
```


## Filter cells and samples.

Suggest removing cells with > 30% mitochondrial UMIs, > 40K UMIs. Consider also dropping samples with median mito UMIs > 30% These libraries would be `1155`, `996`, `1128`, and `898`. Keeping these libraries for now, but should revisit this point. 

```{r}
sobj <- readRDS(file.path("objects", "unfiltered.rds"))

sobj <- subset(sobj, subset = percent.mt < 30)
sobj <- subset(sobj, subset = nCount_RNA < 40000)
sobj <- subset(sobj, subset = nCount_RNA > 500)

#sobj <- subset(sobj, idents = c("1155", "996", "1128", "898"), invert = TRUE)

```


Annotate samples and assign colors to each sample


```{r}

annots <- sobj@meta.data %>% 
  select(UPN, subgroup) %>% 
  unique() %>%
  arrange(subgroup)

subgroup_order <- c(
  "WNT",
  "SHH",
  "GP3",
  "GP3/4",
  "GP4",
  "TBD"
)

annots <- mutate(annots, subgroup = factor(subgroup, levels = subgroup_order))

annots <- annots %>% arrange(subgroup, as.numeric(UPN))

sample_order <- annots$UPN 

sobj@meta.data$UPN <- factor(sobj@meta.data$UPN, levels = sample_order)
sobj@meta.data$subgroup <- factor(sobj@meta.data$subgroup, levels = subgroup_order)


# per sample cols
col_pal <- get_distinct_cols(annots$subgroup)

```


## Normalize and embed into 2D with UMAP

```{r}
sobj <- NormalizeData(sobj, verbose = FALSE)
sobj <- FindVariableFeatures(sobj, 
                             selection.method = "vst", 
                             nfeatures = 3000, 
                             verbose = FALSE)

sobj <- ScaleData(sobj, 
                  vars.to.regress = c("percent.mt", "nCount_RNA"), 
                  verbose = TRUE)

sobj <- RunPCA(sobj, npcs = 40, verbose = FALSE)
sobj <- RunUMAP(sobj, 
                reduction = "pca", 
                dims = 1:40, 
                min.dist = 0.4, 
                n.neighbors = 50L,
                reduction.name = "umap_unaligned",
                reduction.key = "UMAP_")

```

## Markers per unaligned cluster

```{r}

sobj <- FindNeighbors(sobj, 
                      reduction = "pca", 
                      dims = 1:40, 
                      n.neighbors = 50L)

sobj <- FindClusters(sobj, resolution = 0.3)
sobj$clusters_unaligned <- sobj$seurat_clusters
sobj$seurat_clusters <- NULL
sobj$RNA_snn_res.0.3 <- NULL

Idents(sobj) <- "clusters_unaligned"

mkrs <- FindAllMarkers(sobj, only.pos = TRUE)

write_tsv(mkrs, 
          file.path(mkrs_dir, "unaligned_cluster_markers_all_data.tsv"))


```


```{r}
library(harmony)
sobj <- RunHarmony(sobj, "UPN", 
                   plot_convergence = TRUE)
sobj <- RunUMAP(sobj, reduction = "harmony", 
                dims = 1:40, 
                min.dist = 0.4,
                n.neighbors = 50L,
                reduction.name = "umap",
                reduction.key = "UMAP_")

sobj <- FindNeighbors(sobj, reduction = "harmony", dims = 1:40)
sobj <- FindClusters(sobj, resolution = 0.3)

sobj$clusters_aligned <- sobj$seurat_clusters
sobj$seurat_clusters <- NULL
sobj$RNA_snn_res.0.3 <- NULL
Idents(clusters_aligned) <- "clusters_aligned"

mkrs <- FindAllMarkers(sobj, only.pos = TRUE)

write_tsv(mkrs, file.path(mkrs_dir, "aligned_cluster_markers_all_data.tsv"))

```


Write out metadata
```{r} 
mdata_out <- get_metadata(sobj)
outname <- file.path("tables", paste0("metadata_", format(Sys.Date(), "%Y_%m_%d"), ".tsv.gz"))
write_tsv(mdata_out, outname)
```

## save merged object
```{r}
saveRDS(sobj, file.path("objects", "sobj.rds"))
```


```{r}
markers <- read_tsv(file.path(mkrs_dir, "aligned_cluster_markers_all_data.tsv"))

list_of_markers <- split(markers, markers$cluster)
names(list_of_markers) <- str_c("cluster_", names(list_of_markers))
list_of_markers <- map(list_of_markers, 
                       ~set_xlsx_class(.x, "gene", "Text"))

readme_sheet <- data_frame(
  Columns = c(
  "Genes differentially expressed between each cluster and all other cells",
  "",
  "Columns",
  "pval",
  "avg_logFC",
  "pct.1",
  "pct.2",
  "p_val_adj",
  "cluster",
  "gene"
), Description = c(
  "",
  "",
  "",
  "p-value from wilcox test of indicated cluster compared to other clusters",
  "average fold change expressed in natural log",
  "percent of cells expressing gene (UMI > 0) in cluster",
  "percent of cell expressing gene (UMI > 0) in all other clusters",
  "Bonferroni corrected p-value",
  "cluster name",
  "gene name"
))
readme_sheet <- list(README = readme_sheet)
names(readme_sheet) <- "README"
openxlsx::write.xlsx(c(readme_sheet, 
                       list_of_markers),
                     "markers/aligned_cluster_markers.xlsx")
```

```{r}
markers <- read_tsv(file.path(mkrs_dir, "unaligned_cluster_markers_all_data.tsv"))

list_of_markers <- split(markers, markers$cluster)
names(list_of_markers) <- str_c("cluster_", names(list_of_markers))
list_of_markers <- map(list_of_markers, 
                       ~set_xlsx_class(.x, "gene", "Text"))

readme_sheet <- data_frame(
  Columns = c(
  "Genes differentially expressed between each cluster and all other cells",
  "",
  "Columns",
  "pval",
  "avg_logFC",
  "pct.1",
  "pct.2",
  "p_val_adj",
  "cluster",
  "gene"
), Description = c(
  "",
  "",
  "",
  "p-value from wilcox test of indicated cluster compared to other clusters",
  "average fold change expressed in natural log",
  "percent of cells expressing gene (UMI > 0) in cluster",
  "percent of cell expressing gene (UMI > 0) in all other clusters",
  "Bonferroni corrected p-value",
  "cluster name",
  "gene name"
))
readme_sheet <- list(README = readme_sheet)
names(readme_sheet) <- "README"
openxlsx::write.xlsx(c(readme_sheet, 
                       list_of_markers),
                     "markers/unaligned_cluster_markers.xlsx")
```
